src/rpc/mining.cpp

```diff
@@ -1172,21 +1196,32 @@ UniValue submitblock(const UniValue& params, bool fHelp)
     return BIP22ValidationResult(state);
 }
 
-UniValue validateBlock(const UniValue& params, bool fHelp)
+UniValue validateBlocks(const UniValue& params, bool fHelp)
 {
-    const CBlock b2 = blockFromUnivalue(params);
+    std::vector<CBlock> blocks = blocksFromUnivalue(params);
     CValidationState state;
     CBlockIndex* const pindexPrev = chainActive.Tip();
     // TestBlockValidity only supports blocks built on the current Tip
-    uint256 thisPrevBlock = b2.hashPrevBlock;
-    CBlockHeader thisHeader = b2.GetBlockHeader();
+    uint256 thisPrevBlock = blocks[0].hashPrevBlock;
+    CBlockHeader thisHeader = blocks[0].GetBlockHeader();
     uint256 thisHash = thisHeader.GetHash();
     uint256 actualPrevBlock = pindexPrev->GetBlockHash();
     if (thisHash == actualPrevBlock)
         return "i-am-block-producer";
-    if (thisPrevBlock != actualPrevBlock)
+    if (thisPrevBlock != actualPrevBlock) {
+        for (auto desc = begin(blocks) + 1; desc != end(blocks); ++desc) {
+        // if parent block exists and p's prevblick matches actualprevblock
+            if (desc->hashPrevBlock == actualPrevBlock) {
+                // append block in ascending time order and testblockvalidity
+                for (auto asc = desc - 1; asc != begin(blocks); --asc)
```

Comment:
shouldn't the termination condition be `asc != begin(blocks) -1` so we add the transactions from the most recent block?

Reply:
Sharp! Fixed for this too.